styx-scheduler-service/src/main/java/com/spotify/styx/Scheduler.java

```diff
@@ -205,6 +215,64 @@ private void evaluateResourcesForDequeue(
         .map(resource -> ResourceWithInstance.create(resource, instanceState));
   }
 
+  @VisibleForTesting
+  void triggerBackfills(Collection<InstanceState> activeStates) {
+    final List<Backfill> backfills;
+    try {
+      backfills = storage.backfills();
+    } catch (IOException e) {
+      LOG.warn("Failed to get resource limits", e);
+      return;
+    }
+
+    final Map<String, Long> backfillStates = activeStates.stream()
+        // TODO: find out if backfill in other way that string comparison
+        .filter(state -> state.runState().data().triggerId().orElse("").startsWith("backfill-"))
+        .collect(groupingBy(
+            state -> state.runState().data().triggerId().orElse("UNKNOWN"),
+            HashMap::new,
+            counting()));
+
+    backfills.forEach(backfill -> {
+      final Optional<Workflow> workflowOpt =
+          workflowCache.workflow(WorkflowId.create(backfill.component(), backfill.workflow()));
+
+      if (!workflowOpt.isPresent()) {
+        LOG.warn("workflow not found for backfill: {}", backfill);
+        return;
+      }
+
+      final Workflow workflow = workflowOpt.get();
+      final int active = backfillStates.getOrDefault(backfill.id(), 0L).intValue();
+      final int needed = backfill.concurrency() - active;
+      if (needed <= 0) {
+        return;
+      }
+
+      final Instant start = backfill.lastTrigger().orElse(backfill.start());
+      final List<Instant> partitions =
+          ParameterUtil.rangeOfInstants(start, backfill.end(), workflow.schedule().partitioning())
```

Comment:
I think it would be safest to persist the partitioning in the Backfill entity rather than reading it form the workflow. If the partitioning changes during a backfill, this will lead to surprising behaviour.